prophyc/generators/base.py

```diff
@@ -0,0 +1,119 @@
+            language requirements. It's a place to raise exceptions if needed.
+        """
+
+
+class GeneratorBase(GeneratorAbc):
+    def __init__(self, outpu_directory="."):
+        self.output_dir = outpu_directory
+
+    def serialize(self, nodes, base_name):
+        self.check_nodes(nodes)
+
+        for file_path, translator in self.prepare_translators(base_name):
+            file_content = translator(nodes, base_name)
+            _write_file(file_path, file_content)
+
+    def prepare_translators(self, base_name):
+        for extension, translator_class in self.top_level_translators.items():
+            file_path = _make_path(self.output_dir, base_name, extension)
+            yield file_path, translator_class()
+
+
+class BlockTranslatorBase(object):
+    """
+        A translator represents block of content in generated file, (e.g. includes block, constants block, etc..).
+        Translator class can use sub-translator classes as prerequisites (prerequisite_translators).
+        After finishing nodes processing - block_template is applied on generated content (block_post_process).
+
+        To enable translation of given node type just implement a corresponding method from
+        translation_methods_map. The method has to take a single node and return a string with translation result.
+
+        Dispatcher will skip translation of nodes of types that have no translation implemented.
+    """
+    translation_methods_map = {
+        model.Constant: "translate_constant",
+        model.Enum: "translate_enum",
+        model.Include: "translate_include",
+        model.Struct: "translate_struct",
+        model.Typedef: "translate_typedef",
+        model.Union: "translate_union",
+    }
+    prerequisite_translators = []
+    block_template = None
+
+    def __call__(self, nodes, base_name):
+        content = self.process_nodes(nodes, base_name)
+        content = self.block_post_process(content, base_name, nodes)
+        return content
+
+    def process_nodes(self, nodes, base_name):
+        return ''.join(self._nodes_dispatcher(nodes, base_name))
+
+    @classmethod
+    def block_post_process(cls, content, base_name, nodes):
+        if cls.block_template:
+            return cls.block_template.format(content=content, base_name=base_name, nodes=nodes)
+        else:
+            return content
+
+    @classmethod
+    def prepend_newline(cls, previous_node, current_node):
+        is_different_type = type(previous_node) is not type(current_node)
+        is_enum_struct_or_union = isinstance(previous_node, (model.Enum, model.Struct, model.Union))
+
+        if previous_node:
+            return is_different_type or is_enum_struct_or_union
+
+    def _nodes_dispatcher(self, nodes, base_name):
+        for block_translator_class in self.prerequisite_translators:
+            prerequisite_block_translator = block_translator_class()
+            yield prerequisite_block_translator(nodes, base_name)
```

Comment:
There could be a newline between prerequisite blocks by default, then user (derived class) wouldn't have to worry about adding such newlines on its own. Of course there would have to be logic similar to newlines between nodes here: if previous prerequisite_translator block is non-empty, then before appending next prerequisite_translator - of first node - newline has to be added.

If cost of doing so is very large - don't bother.